shared/src/ledger/eth_bridge/bridge_pool_vp.rs

```diff
@@ -0,0 +1,555 @@
+        }
+    }
+}
+
+impl<'a, D, H, CA> NativeVp for BridgePoolVp<'a, D, H, CA>
+where
+    D: 'static + DB + for<'iter> DBIter<'iter>,
+    H: 'static + StorageHasher,
+    CA: 'static + WasmCacheAccess,
+{
+    type Error = Error;
+
+    const ADDR: InternalAddress = InternalAddress::EthBridgePool;
+
+    fn validate_tx(
+        &self,
+        tx_data: &[u8],
+        keys_changed: &BTreeSet<Key>,
+        _verifiers: &BTreeSet<Address>,
+    ) -> Result<bool, Error> {
+        tracing::debug!(
+            tx_data_len = tx_data.len(),
+            keys_changed_len = keys_changed.len(),
+            verifiers_len = _verifiers.len(),
+            "Ethereum Bridge Pool VP triggered",
+        );
+        let signed: SignedTxData = BorshDeserialize::try_from_slice(tx_data)
+            .map_err(|e| Error(e.into()))?;
+
+        let transfer: PendingTransfer = match signed.data {
+            Some(data) => BorshDeserialize::try_from_slice(data.as_slice())
+                .map_err(|e| Error(e.into()))?,
+            None => return Ok(false),
+        };
+
+        // check that the signed root is not modified
+        let signed_root_key = get_signed_root_key();
+        if keys_changed.contains(&signed_root_key) {
+            return Ok(false);
+        }
+        // check that the pending transfer (and only that) was added to the pool
+        // TODO: This will change slightly when we merkelize the pool,
+        // but that will be a separate PR.
+        let pending_key = get_pending_key();
+        let pending_pre: HashSet<PendingTransfer> =
+            self.ctx.read_pre_value(&pending_key).ok_or(eyre!(
+                "The bridge pool transfers are missing from storage"
+            ))?;
+        let pending_post: HashSet<PendingTransfer> =
+            self.ctx.read_post_value(&pending_key).ok_or(eyre!(
+                "The bridge pool transfers are missing from storage"
+            ))?;
+        if !pending_post.contains(&transfer) {
+            return Ok(false);
+        }
+        for item in pending_pre.symmetric_difference(&pending_post) {
+            if item != &transfer {
+                return Ok(false);
+            }
+        }
+
+        // check that gas fees were put into escrow
+
+        // check that the correct amount was deducted from the fee payer
+        if let Some(SignedAmount::Negative(amount)) =
+            self.account_balance_delta(&transfer.gas_fee.payer)
+        {
+            if amount != transfer.gas_fee.amount {
+                return Ok(false);
+            }
+        } else {
+            return Ok(false);
+        }
+        // check that the correct amount was credited to escrow
+        if let Some(SignedAmount::Positive(amount)) =
+            self.account_balance_delta(&BRIDGE_POOL_ADDRESS)
+        {
+            if amount != transfer.gas_fee.amount {
+                return Ok(false);
+            }
+        } else {
+            return Ok(false);
+        }
```

Comment:
I think we should also explicitly `match` rather than using `if let Some(...)` so as to cover if `self.account_balance_delta` returns `None` . In case reading a value fails for whatever reason, we will want to reject the transaction (or return an error maybe?).

Reply:
It does reject, it is an if let Some { ..} else { .. } block. I can add logging though.

Reply:
> It does reject, it is an if let Some { ..} else { .. } block. I can add logging though.

oops, completely missed that!